servicemanager/servicemanager_test.go

```diff
@@ -522,169 +539,431 @@ func TestRemoveService(t *testing.T) {
 		LayersDir:      path.Join(tmpDir, "layers"),
 		DownloadDir:    path.Join(tmpDir, "downloads"),
 		ServiceTTLDays: 2,
 	}
 
-	layerProvider := &testLayerProvider{
-		availableSize:  19 * 1024,
-		allocationSize: 20 * 1024,
-	}
-
-	servicemanager.GetAvailableSize = layerProvider.getAvailableSize
+	layerProvider := &testLayerProvider{}
 
-	sm, err := servicemanager.New(config, layerProvider, serviceStorage)
+	sm, err := servicemanager.New(config, serviceStorage, layerProvider)
 	if err != nil {
 		t.Fatalf("Can't create SM: %v", err)
 	}
 
-	serviceID := "testAllocateMemory"
+	serviceID := "testRestoreService"
 
-	serviceURL, fileInfo, err := prepareService("Service content")
+	serviceURL, fileInfo, layerDigest, err := prepareService("Service content", defaultServiceSize)
 	if err != nil {
 		t.Fatalf("Can't prepare test service: %v", err)
 	}
 
+	layerProvider.digests = layerDigest
+
 	serviceInfo := servicemanager.ServiceInfo{ServiceID: serviceID, AosVersion: 1}
 
 	if err = sm.InstallService(serviceInfo, serviceURL, fileInfo); err != nil {
 		t.Fatalf("Can't install service: %v", err)
 	}
 
-	if err := sm.ApplyService(serviceInfo); err != nil {
-		t.Fatalf("Can't apply service: %v", err)
+	if err = sm.RestoreService(serviceInfo.ServiceID); err != nil {
+		t.Fatal("Can't restore service")
 	}
 
 	if err := sm.UseService(serviceInfo.ServiceID, serviceInfo.AosVersion); err != nil {
 		t.Fatalf("Can't set use service: %v", err)
 	}
 
+	service, err := sm.GetServiceInfo(serviceID)
+	if err != nil {
+		t.Fatalf("Can't get service info: %v", err)
+	}
+
+	if service.Cached {
+		t.Error("Service should not be cached")
+	}
+
+	serviceStorage.cachedServiceError = true
+
+	if err = sm.RemoveService(serviceID); err == nil {
+		t.Fatal("Should be error to set cached service")
+	}
+
+	serviceStorage.cachedServiceError = false
+
 	if err = sm.RemoveService(serviceID); err != nil {
 		t.Fatalf("Can't remove service: %v", err)
 	}
 
-	serviceInfo, err = sm.GetServiceInfo(serviceID)
+	if err = sm.InstallService(serviceInfo, serviceURL, fileInfo); err == nil {
+		t.Fatal("Should be error service cashed")
+	}
+
+	service, err = sm.GetServiceInfo(serviceID)
+	if err != nil {
+		t.Fatalf("Can't get service info: %v", err)
+	}
+
+	if !service.Cached {
+		t.Error("Service should be cached")
+	}
+
+	serviceStorage.cachedServiceError = true
+
+	if err = sm.RestoreService(serviceInfo.ServiceID); err == nil {
+		t.Fatal("Should be error to set cached service")
+	}
+
+	serviceStorage.cachedServiceError = false
+
+	if err = sm.RestoreService(serviceInfo.ServiceID); err != nil {
+		t.Fatalf("Can't remove service: %v", err)
+	}
+
+	service, err = sm.GetServiceInfo(serviceID)
 	if err != nil {
 		t.Fatalf("Can't get service info: %v", err)
 	}
 
-	if serviceInfo.AosVersion != 1 {
-		t.Errorf("Unexpected aos version: %v", serviceInfo.AosVersion)
+	if service.Cached {
+		t.Error("Service should not be cached")
+	}
+}
+
+func TestAllocateMemoryInstallService(t *testing.T) {
```

Comment:
Mock space allocator in order to do not mount/umount test disk.

Reply:
Done